src/Std/Sync/Context.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+module
+
+prelude
+public import Std.Data
+public import Init.System.Promise
+public import Init.Data.Queue
+public import Std.Sync.Mutex
+public import Std.Sync.CancellationToken
+public import Std.Internal.Async.Select
+
+public section
+
+/-!
+This module provides a tree-structured cancellation context called `Context` where cancelling a parent
+automatically cancels all child contexts.
+-/
+
+namespace Std
+open Std.Internal.IO.Async
+
+/--
+The central state structure shared by all context types.
+-/
+structure ContextState where
+  /--
+  Map of token IDs to optional tokens and their children.
+  `none` represents a background context that cannot be cancelled.
+  -/
+  tokens : TreeMap UInt64 (Option CancellationToken × Array UInt64) := .empty
+
+  /--
+  Next available ID
+  -/
+  id : UInt64 := 1
+
+/--
+A cancellation context that allows multiple consumers to wait until cancellation is requested. Forms
+a tree structure where cancelling a parent cancels all children.
+-/
+structure Context where
+  state : Std.Mutex ContextState
+  token : CancellationToken
+  id : UInt64
+
+namespace Context
+
+/--
+Creates a new root cancellation context.
+-/
+def new : BaseIO Context := do
+  let token ← Std.CancellationToken.new
+  return {
+    state := ← Std.Mutex.new { tokens := .empty |>.insert 0 (some token, #[]) },
+    token,
+    id := 0
+  }
+
+/--
+Forks a child context from a parent. If the parent is already cancelled, returns the parent context.
+Otherwise, creates a new child that will be cancelled when the parent is cancelled.
+-/
+def fork (root : Context) : BaseIO Context := do
+  if ← root.token.isCancelled then
+    return root
+
+  let token ← Std.CancellationToken.new
```

Comment:
I believe that there is a race condition here. If the root token is cancelled after the check in the beginning of `fork` but before the critical section below is entered, then you can have a context where the parent is cancelled but the child is not.

Reply:
Fixed